torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -204,6 +207,10 @@ def _init_mp_dtypes(self) -> None:
 
     def lazy_init(self):
         # Lazy init should be idempotent
+        if self.is_sharded and not self._reset_sharded_params:
```

Comment:
@weifengpy 
1. I do not think we need `torch.no_grad` context since `reset_sharded_param` operates on the local tensor that does not require grad.
2. I think we want this `self._reset_sharded_params` bool to make this only run once because otherwise every `FSDPModule.unshard()` call will loop over the sharded parameters and copy them. It is hard to tell whether `reset_sharded_param` should be a no-op or not, which is why we cannot make it short-circuit if it should be a no-op.